keps/sig-storage/3141-prevent-volume-mode-conversion/README.md

```diff
@@ -0,0 +1,734 @@
+
+- With feature flag disabled:
+  - attempt to convert volume mode when creating a `PVC`
+  from a `VolumeSnapshot`.
+- With feature flag enabled, attempt to convert volume mode when creating a `PVC`
+from a `VolumeSnapshot`:
+  - With `Spec.SourceVolumeMode` populated and `snapshot.storage.kubernetes.io/allowVolumeModeChange: true`
+  annotation present.
+  - With `Spec.SourceVolumeMode` populated but no `snapshot.storage.kubernetes.io/allowVolumeModeChange: true`
+  annotation.
+  - With `Spec.SourceVolumeMode` set to `nil`.
+
+#### E2E tests
+
+The feature flag will be enabled for e2e tests. The tests will attempt to convert volume 
+mode when creating a `PVC` from a `VolumeSnapshot`:
+  - With `Spec.SourceVolumeMode` populated and `snapshot.storage.kubernetes.io/allowVolumeModeChange: true`
+    annotation present.
+  - With `Spec.SourceVolumeMode` populated but no `snapshot.storage.kubernetes.io/allowVolumeModeChange: true`
+    annotation.
+  - With `Spec.SourceVolumeMode` set to `nil`.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind an out-of-tree feature flag.
+- Feedback from users.
+- Implementation of unit and e2e tests.
+
+#### Alpha -> Beta
+
+- One release with positive feedback from users. 
+
+#### Beta -> GA
+
+- Deployed in production and in use by backup software. 
+- Gone through one kubernetes upgrade.
+
+### Upgrade / Downgrade Strategy
+
+1. Upgrading `external-snapshotter` and `external-provisioner` with `PreventVolumeModeConversion`
+enabled:
+- `VolumeSnapshots` created after the upgrade will maintain a reference to the
+source volume mode. Newly created `PVCs` will undergo an additional check before the 
+provisioning is performed on the storage backend. 
+- `VolumeSnapshots` created before the upgrade will leave the new API field 
+unpopulated.
+
+2. Downgrading `external-snapshotter` and `external-provisioner` with `PreventVolumeModeConversion`
+disabled: 
+- `VolumeSnapshots` created prior to the upgrade will still maintain a reference 
+to the source volume mode, but `PVCs` can be created from them without the
+additional check. 
+
+### Version Skew Strategy
+
+This proposal requires changes to three components - `VolumeSnapshotContent` API, 
+`external-snapshotter` and `external-provisioner`. 
+
+If any of the components are not upgraded to a version supporting this feature,
+then the feature will not work as expected. From an end user perspective, the
+existing behavior will continue, ie, there will be no check to prevent 
+unauthorized conversion of the volume mode. 
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Other
+  - Describe the mechanism: Out-of-tree flag named `PreventVolumeModeConversion`, 
+  which will be enabled in `external-provisioner` and `external-snapshotter`.
+  - Will enabling / disabling the feature require downtime of the control
+    plane? No
```

Comment:
If we change `PreventVolumeModeConversion` flag from false to true or from true to false, we need to restart `external-provisioner` and `external-snapshotter`, right?

Reply:
Yes, added that